### ISSUE TITLE
GNOME: Add post_install() script

### ISSUE THREAD

**Issue by xclaesse** [mesonbuild/meson#8268]
Typical GNOME application must update schema/icon/giomodule cache in the final destination directory. See for example:
- https://gitlab.gnome.org/GNOME/gtk/-/blob/master/build-aux/meson/post-install.py
- https://gitlab.gnome.org/GNOME/gedit/-/blob/master/build-aux/meson/post_install.py

That script is less trivial than it seems, those binaries could come from subproject, or pkg-config variable but pkg-config binary could be missing on Windows, or in system $PATH. In the case glib/gtk are subprojects we need to run the installed binary and not the one overriden in build directory, otherwise on Windows we have to deal with DLL location in PATH.

**Comment by xclaesse** [comment]
See installation error of gtk master on Windows: https://gitlab.gnome.org/GNOME/gtk/-/issues/3626.

**Comment by ebassi** [comment]
The problem is that you only want to run `glib-compile-schemas` and friends at the *end* of the installation process of *all* dependencies of a project. That's why all those scripts check for the `DESTDIR` environment variable being set. We definitely don't want this to run every time we install GTK as part of the normal packaging process, or when using tools like jhbuild or flatpak.

Additionally, the set of tools being called depend heavily on the project, and in some cases might be installed by the project itself; for instance, gdk-pixbuf will call `gdk-pixbuf-query-loaders`:

 - https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/build-aux/post-install.sh

**Comment by xclaesse** [comment]
> The problem is that you only want to run `glib-compile-schemas` and friends at the _end_ of the installation process of _all_ dependencies of a project. That's why all those scripts check for the `DESTDIR` environment variable being set. We definitely don't want this to run every time we install GTK as part of the normal packaging process, or when using tools like jhbuild or flatpak.

Sure, we should still check for DESTDIR. More generally it would actually make sense for `meson.add_install_script()` to have a keyword argument telling to not run it in the case DESTDIR is set.

> Additionally, the set of tools being called depend heavily on the project, and in some cases might be installed by the project itself; for instance, gdk-pixbuf will call `gdk-pixbuf-query-loaders`:
> 
>     * https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/build-aux/post-install.sh

Could be ported to cross platform python instead of separate bat/sh.[/off-topic] That script seems specific to gdk-pixbuf, it can remain there. Howerver it comforts me in the idea that we need to add `meson.add_install_script('foo.py', skip_if_destdir: true)` because that way meson could print a generic message at the end listing all scripts that have been skipped.

Of course each app needs different cache updates, what I had in mind is something like `gnome.post_install(glib_schemas: true, gtk4_icon_update: true)`. It will run only the parts requested, but collect from every subproject, so maybe another subproject will do `gnome.post_install(gio_modules: true)` and that will in the end run all 3 scripts, and only if DESTDIR is not set.

**Comment by xclaesse** [comment]
Let's start with that: https://github.com/mesonbuild/meson/pull/8269.

**Comment by Neumann-A** [comment]
since I am working on https://github.com/microsoft/vcpkg/pull/13100

> end of the installation process of all dependencies of a project. 

Why? Is there something wrong doing it per project? Or is it just an optimization?

**Comment by ebassi** [comment]
> Why? Is there something wrong doing it per project? Or is it just an optimization?

It depends.

If a tool is directly installing in the target directory, then calling a script to rebuild a bunch of different cache files once is just an optimisation.

If a tool is installing a project into a separate `DESTDIR`, and then moving the files to the intended target directory (for instance, to keep inodes "near" each other via hardlink/removal, or to post-process the installation by removing libtool's `*.la` files), then running the script for each project after the installation is going to produce the wrong result, and you really want to do that at the end of the entire installation.

Packaging tools such as RPM or dpkg, as well as tools like [jhbuild](https://wiki.gnome.org/Projects/Jhbuild), have had this kind of "trigger" for years.

**Comment by Neumann-A** [comment]
> script for each project after the installation is going to produce the wrong result

Why or in what regard? What exactly is wrong with the results? What will/can happen?

>post-process the installation
 
To clarify vcpkg does do post-process by using a staged install where all build results are installed into a different directory (staging area) and later moved/copied to the final installed tree. This is done to remove e.g. libtool files, correct pc/cmake files, check correct architecture of installed libs etc.

Having to run some script in the final install directory seems to be a code smell in my eyes since it potentially can break the installed directory completely and is a sign of insufficient modularization/separation.

**Comment by ebassi** [comment]
> Why or in what regard? What exactly is wrong with the results? What will/can happen?

If a tool installs things under `$DESTDIR/$prefix` and you run the script *before* the tool moves the files from out `$DESTDIR`, then any cache file that contains paths will reference `$DESTDIR/$prefix` instead of just `$prefix`. This is why all scripts bail out if `$DESTDIR` is set.

> Having to run some script in the final install directory seems to be a code smell in my eyes since it potentially can break the installed directory completely and is a sign of insufficient modularization/separation.

That'd pretty much invalidate things like creating memory mappable binary blobs out of installed files like settings schemas, desktop files, or out-of-tree modules that extend core functionality.

I understand that it looks weird from the perspective of a Windows developer, with every application existing in a self-contained prefix and with *very* minimal shared components; but on OSes where the file systems structure defines a shared namespace these things matter a lot more.

**Comment by Neumann-A** [comment]
> cache file that contains paths will reference $DESTDIR/$prefix instead of just $prefix

> I understand that it looks weird from the perspective of a Windows developer, with every application existing in a self-contained prefix and with very minimal shared components; but on OSes where the file systems structure defines a shared namespace these things matter a lot more.

If `<things>` would stop assuming absolute paths and just start using relative paths that would not be a problem and would make so many things much simpler in regards to relocating of  `<things>`. For example the pkg-config files in vcpkg all get changed to use `prefix=../../${pcfiledir}` (If installed into `prefix/lib/pkgconfig`)

**Comment by ebassi** [comment]
> If `<things>` would stop assuming absolute paths and just start using relative paths that would not be a problem and would make so many things much simpler in regards to relocating of `<things>`.

Again, that might make sense on Windows, but not everything has to be relocatable—especially if it's part of the core OS.

All the cache files I'm referring to are literally system components: settings schemas, desktop file caches, icon theme caches, image loading modules, extension modules for libraries, etc. Some of them can be installed by applications, but there's literally no point in saying "the desktop file cache in `/usr/share/applications` is relative to something else" (who knows what); or: "the image loading modules for libgdk-pixbuf-2.0.so.0 are relative to the shared library in `"../../usr/lib64"`", if they are also in `/usr/lib64`.

Anyway, this is a weird tangent, so we should stop it here; projects launching these tools will continue to bail out when `DESTDIR` is set because that's how packaging on Linux (and other Unix-like OSes) works.

**Comment by xclaesse** [comment]
Take for example `gtk-update-icon-cache`, it creates an index of all icons installed in a directory. You could have `app1` that is already installed into its final destination (e.g. `/usr` prefix) and has `icon1.png` that is referenced in `/usr/share/icon/hicolor/index.theme`.

Now you build a new `app2` that contains `icon2.png` and install it in a temporary DESTDIR (e.g. `/home/user/usr`). If you run gtk-update-icon-cache at that point, it will create the file `/home/user/usr/share/icons/hicolor/index.theme` referencing only `icon2.png` because it's the only icon it found in that directory.

When you copy all that tree into the final destination, your `index.theme` that refer only `icon2.png` overrides the one that refer `icon1.png`. To fix that problem, you should not run `gtk-icon-update-cache` when installing into a DESTDIR, it is the tool that copy into final destination that is responsible to run `gtk-icon-update-cache` in `/usr/share/icons/hicolor` after `icon2.png` has been copied there, so both `icon1.png` and `icon2.png` gets referenced into the cache.

Since the copy from DESTDIR to final destination is out of scope for Meson, it is up to vcpkg/dpkg/rpm to do it. As an extra optimization, they run `gtk-update-icon-cache` only once after installing multiple applications instead of running it after each package. In dpkg that's called "triggers".

**Comment by eli-schwartz** [comment]
> Why? Is there something wrong doing it per project? Or is it just an optimization?

Every time you run it, it completely discards the output of the previous run and rebuilds the index from scratch. Then people moving from a staging directory to a final directory need to delete it anyway, because hundreds of projects all write to the same global caching database.

This is the case for the gtk icon cache as well as all other programs that are typically moved into a post-install hook or trigger or similar Linux distro packaging mechanism.

To put it into terms a windows user would understand, your software doesn't install its own Ntuser.dat in order to deploy registry keys. Nor does it run a command that creates a Ntuser.dat containing registry keys just for your software, then expect vcpkg to copy that over your user profile and overwrite everything that used to be there.

Every program registering a Windows registry key needs to **"play together*** and likewise for the gtk icon cache. It is, after all, a cache -- not the actual icon file.

**Comment by eli-schwartz** [comment]
Although this PR doesn't implement anything other than gtk stuff, the same rules apply to freedesktop `update-desktop-database` and `update-mime-database`.

**Comment by xclaesse** [comment]
> Although this PR doesn't implement anything other than gtk stuff, the same rules apply to freedesktop `update-desktop-database` and `update-mime-database`.

I have kept them out of scope for now because I think it belongs to the future freedesktop module: https://github.com/mesonbuild/meson/pull/6971. But indeed those 2 are good candidate for a fdo.post_install() method there.
